### PR TITLE
Big-step version of `SYNTH`

### PR DESCRIPTION
**Pull request by jonsterling**
This removes the individual `SYNTH` rules and replaces them with a single big-step synthetic rule. This makes the proof search process a bit easier to control, and is connected with @cangiuli and my Thought that we may wish to move toward big-step synthetic style rules in many places (considering the massive performance increase that we achieved by using a synthetic rule for applications). It's also related to #484.

Note that this change does not appear to significantly improve performance.

@favonia I'm curious about your feedback; we're not insisting on making this change, but it seemed like something worth trying out. What do you think about it?

### CODE REVIEW

**Comment by favonia** [comment]
I am a practical person who is fine with whatever works and can potentially be proved correct.

**Comment by favonia** [comment]
Any objection? I would like to resolve #559 but that would make more sense after this PR.

**Comment by jonsterling** [comment]
@favonia I think it should be OK; but I'd like to have one last convo with @cangiuli about it before merging.

**Comment by jonsterling** [comment]
P.S. Somewhere I have some code that combines this with #559, which I tried out on the airplane. It may make sense to do these both simultaneously, especially if I can revive that code.

**Comment by favonia** [comment]
That would be great! Still, let me make the PR just in case.

**Comment by jonsterling** [comment]
I'm going to take on the task of bringing this PR up to date, and restoring the ideas I had some weeks ago for improving `auto` in the presence of the changes proposed here. Then, I'll fix the semi-simplicial example (hopefully).

**Comment by favonia** [comment]
Oops I just did it. (BTW I intended to assign myself to make it clear.)

**Comment by favonia** [comment]
Feel free to do whatever ugly git tricks to make this PR great again.

**Comment by jonsterling** [comment]
@favonia OK, no problem! I've just written a ton of code to get this back into shape (since the PR was very far away from done), so if you don't mind, I want to force-push.

**Comment by jonsterling** [comment]
Actually, I'll just open a new PR.

**Comment by jonsterling** [comment]
Actually, it wasn't that bad to just merge the merge into the merge 😄

**Comment by jonsterling** [comment]
Alright, I've unloaded my Thought into this branch. Performance is a little bit worse across the board, but not devastatingly so (though I'd be interested to find out why--it may have to do with the fact that I made `auto` a little more aggressive in applying invertible elim rules). On the bright side, the semi-simplicial example is fairly trivial now.

**Comment by favonia** [comment]
Then we can merge the merge that was the merge of something that had been merged into some merge. Did I count correctly?

**Comment by jonsterling** [comment]
To clarify the content of this PR, it removes the `SYNTH, MATCH, MATCH_RECORD` judgments, and replaces them with a "synchronous" helper function that generates a type together with whatever subgoals are necessary for the input to be in that type. Then, this helper function is called in many of the equality rules, and direct pattern matching replaces the `MATCH` subgoals, etc.

This makes many of the equality rules "synthetic"...

**Comment by jonsterling** [comment]
This resolves the problem where both `Path.AppEq` and `Line.AppEq` were applicable, but only one of them would ultimately work. Previously we tried to deal with this using backtracking, but that was really devastating (blowup!)... Now, these rules fail synchronously when `synthTerm` doesn't give what is expected.

The tradeoff is that the user must get the goal into the right shape prior to applying these rules. Usually this is just a matter of making sure it is neutral (which can be done by applying computation rules). Of course, all this is dealt with in our automation.

**Comment by cangiuli** [comment]
I'll keep tinkering with `examples/groupoid.prl` in the future: it's kind of arbitrary what I chose to prove, and there's a bit of code duplication. But these proofs were incredibly frustrating due to backtracking before this PR, and now they're more or less immediate by `exact`.

**Comment by favonia** [comment]
So the solution to improve searching is to not search at all!

**Comment by favonia** [comment]
Is there any reason not to merge this PR?

File: example/groupoid.prl [hunk 1]
```diff
@@ -66,9 +66,7 @@ Thm Trans/Unit/R(#l:lvl) : [
   refine path/eq/from-line; auto
 ].
 
-
-/* TODO
- 
+// Thanks to Bruno Bentzen for already doing this and drawing a picture! -Carlo
```
[the remark this answers is not in the document]
Comment (cangiuli) [reply example/groupoid.prl]: We can work on the neural interface once `auto` is fixed.
cube

[the remark this answers is not in the document]
Comment (favonia) [reply example/groupoid.prl]: LOL

[the remark this answers is not in the document]
Comment (jonsterling) [reply example/groupoid.prl]: omfg

[the remark this answers is not in the document]
Comment (jonsterling) [reply example/groupoid.prl]: send this to bob

**Comment by jonsterling** [comment]
I vote we merge!

**Pull request merged**